src/language/schema/materializer.js

```diff
@@ -0,0 +1,263 @@
+  array.forEach((item) => {
+    map[keyFn(item)] = valueFn(item);
+  });
+  return map;
+}
+
+function buildWrappedType(innerType, inputTypeAST) {
+  if (inputTypeAST.kind === LIST_TYPE) {
+    return new GraphQLList(buildWrappedType(innerType, inputTypeAST.type));
+  }
+  if (inputTypeAST.kind === NON_NULL_TYPE) {
+    return new GraphQLNonNull(buildWrappedType(innerType, inputTypeAST.type));
+  }
+  return innerType;
+}
+
+function getInnerTypeName(typeAST) {
+  if (typeAST.kind === LIST_TYPE || typeAST.kind === NON_NULL_TYPE) {
+    return getInnerTypeName(typeAST.type);
+  }
+  return typeAST.name.value;
+}
+
+/**
+ * This takes the ast of a schema document produced by parseSchema in
+ * src/language/schema/parser.js.
+ *
+ * Given that AST it constructs a GraphQLSchema. As constructed
+ * they are not particularly useful for non-introspection queries
+ * since they have no resolve methods.
+ */
+export function materializeSchema(ast, queryTypeName, mutationTypeName) {
+  if (nullish(ast)) {
+    throw new Error('must pass in ast');
+  }
+  if (nullish(queryTypeName)) {
+    throw new Error('must pass in query type');
+  }
+
+  var astMap = makeDict(ast.definitions, d => d.name.value);
+
+  /**
+   * This generates a function that allows you to produce
+   * type definitions on demand. We produce the function
+   * in order to close over the memoization dictionaries
+   * that need to be retained over multiple functions calls.
+   **/
+  function getTypeDefProducer() {
+
+    var innerTypeMap = {
+      String: GraphQLString,
+      Int: GraphQLInt,
+      Boolean: GraphQLBoolean,
+      ID: GraphQLID,
+    };
+
+    return (typeAST) => {
+      var typeName = getInnerTypeName(typeAST);
+      if (!nullish(innerTypeMap[typeName])) {
+        return buildWrappedType(innerTypeMap[typeName], typeAST);
+      }
+
+      if (nullish(astMap[typeName])) {
+        throw new Error(`Type ${typeName} not found in document`);
+      }
+
+      var innerTypeDef = makeSchemaDef(astMap[typeName]);
+      if (nullish(innerTypeDef)) {
+        throw new Error('Nothing constructed for ' + typeName);
+      }
+      innerTypeMap[typeName] = innerTypeDef;
+      return buildWrappedType(innerTypeDef, typeAST);
+    };
+  }
+
+
+  var produceTypeDef = getTypeDefProducer(ast);
+
+  if (nullish(astMap[queryTypeName])) {
+    throw new Error(`Type ${queryTypeName} not found in document`);
+  }
+
+  var queryType = produceTypeDef(astMap[queryTypeName]);
+  var schema;
+  if (nullish(mutationTypeName)) {
+    schema = new GraphQLSchema({query: queryType});
+  } else {
+    schema = new GraphQLSchema({
+      query: queryTypeName,
+      mutation: produceTypeDef(astMap[mutationTypeName]),
+    });
+  }
+
+  // This actually constructs all the types by iterating over the schema
+  // This makes it so that errors actually get caught before this function
+  // exits.
+  schema.getTypeMap();
+  return schema;
+
+  function makeSchemaDef(def) {
```

Comment:
It's so they can have access to the memoization maps without having to thread them around.